interpreter/cling/lib/Interpreter/Interpreter.cpp

```diff
@@ -463,8 +463,8 @@ namespace cling {
     const char* Attr = "";
 #endif
 
-#if defined(__GLIBCXX__)
-    const char* cxa_atexit_is_noexcept = LangOpts.CPlusPlus ? " noexcept" : "";
+#if defined(__GLIBCXX__ || _LIBCPP_VERSION)
+    const char* cxa_atexit_is_noexcept = noexcept(__cxa_atexit) ? " noexcept" : "";
 #else
     const char* cxa_atexit_is_noexcept = "";
 #endif
```

Comment:
Wouldn't just:
```suggestion
    const char* cxa_atexit_is_noexcept = noexcept(__cxa_atexit) ? " noexcept" : "";
```
also work?

But then again the new test (is the `cxa_atexit` seen during compilation of this source file) and the old test (whether the interpreter is in C++ mode or not) are seemingly completely different.

And given that commit log 2a3066141e10a0280f90a0ef13a3a5452b5ce43a says:
```
However, since the
    __cxa_atexit function is declared as extern "C" in libc++, we cannot
    add "noexcept" to it, as is done for libstdc+
```
It might indeed by `libc++` version dependent, so maybe we need:
```
    const char* cxa_atexit_is_noexcept = (noexcept(__cxa_atexit) && LangOpts.CPlusPlus) ? " noexcept" : "";
```
but then again 1ce69935d0ffa29f1f6fcc46020839d7b3367793 did not clarify why the `LangOpts.CPlusPlus` was needed.

Reply:
> Wouldn't just ... also work?

`__cxa_atexit` doesn't exist on MSVC so `noexcept(__cxa_atexit)` isn't a thing.

Reply:
> But then again the new test (is the cxa_atexit seen during compilation of this source file) and the old test (whether the interpreter is in C++ mode or not) are seemingly completely different.

We require compile-time == runtime, else the symbols will differ, too.

Reply:
> And given that commit log https://github.com/root-project/root/commit/2a3066141e10a0280f90a0ef13a3a5452b5ce43a says:

I was hoping that this is addressed by the current `noexcept(__cxa_atexit)` - but it isn't: `extern "C"` means `noexcept`, but you cannot redeclare the function as `noexcept` because it's `extern "C"`. Argh. OK, back to the drawing board.

Reply:
> __cxa_atexit doesn't exist on MSVC so noexcept(__cxa_atexit) isn't a thing.

Should we define it for windows then?  I.e. does the code around line 487 have the "desired" side effect on Windows?

Reply:
> We require compile-time == runtime, else the symbols will differ, too.

Not strictly (`extern "C"` but also CUDA mode)

Reply:
> Argh. OK, back to the drawing board.

👍